### DISCUSSION TITLE
use cue to vet another cue file

### DISCUSSION THREAD

**Discussion by sahroshan** [cue-lang/cue#2211] [Q&A] [answered]
Hi All,
I have been trying to use  cue to valiidate another cue. Cuelang report errors for defined keys. But not when there is an extra key at root level.  Below are the steps to reproduce it.

### steps to reproduce
1. create a schema file
```
-- schema.cue
#Config: close({
	foo: close({
		bar: string | *"default from schema"
	})
})

#Config
```

2. create simple cue file
```
-- sub.cue
extra: "this extra field"
```
3. run vet command `cue vet sub.cue schema.cue`

### expected
error `extra` field found in cue

### got

cue vet exits without any error

Am i doing something wrong here? Please suggest possible fixes or workarounds

**Comment by verdverm** [comment]
The root value is not closed, though that is being reconsidered.

The easiest thing to do for now is put your full schema in a definition and use the `-e` to select it for use during validation

**Reply by sahroshan** [reply]
Thank you for reply. I am able to reproduce the issue with closed root.

```
-- schema.cue
close({
	foo: {
		bar: string
	}
})

```

```
-- invalid.cue
foo: {
	bar: "this is valid field"
}
extra: "this extra field"
```

# eval output
```
➜  cue eval invalid.cue schema.cue 
foo: {
    bar: "this is valid field"
}
extra: "this extra field"
```

# vet also does not raise  any  error for extra field in invalid.cue

**Reply by verdverm** [reply]
That does not close the root value. There is no way to do so currently. It requires an implementation change

**Reply by sahroshan** [reply]
Thank you for the reply.

Would you know  if  implementing closed root is in cue roadmap?

**Reply by verdverm** [reply]
I don't know. Maybe @mpvl or someone else from the CUE team can comment.

**Reply by mpvl** [reply]
The top-level thing not closing is only relevant for applying a schema within the package itself. We are exploring various solutions for this. But that is not necessary for solving this issue.

**Reply by sahroshan** [reply]
Thank you @mpvl for the clarifications.

**Comment by mpvl** [comment, marked as answer]
This is a bug, really. Two .cue files without a package clause should be considered as separate files, and thus the embedded definition should close it and this should work. We know where the problem is but haven't fixed it yet.

A workaround for now is to make schema.cue a separate package and use it as such:
```
! exec cue vet :schema sub.cue
cmp stderr expect-stderr
-- schema.cue --
package schema
#Config: {
	foo: {
			bar: string | *"default from schema"
	}
}
#Config

-- sub.cue --
extra: "this extra field"

-- expect-stderr --
extra: field not allowed:
    ./schema.cue:1:1
    ./schema.cue:2:10
    ./schema.cue:7:1
    ./sub.cue:1:1
```

Because of the same bug, it currently doesn't work to use `cue vet schema.cue sub.cue`, but using `:schema` works.
